common/src/unifycr_keyval.c

```diff
@@ -0,0 +1,673 @@
+/*
+ * Copyright (c) 2019, Lawrence Livermore National Security, LLC.
+ * Produced at the Lawrence Livermore National Laboratory.
+ *
+ * Copyright 2019, UT-Battelle, LLC.
+ *
+ * LLNL-CODE-741539
+ * All rights reserved.
+ *
+ * This is the license for UnifyCR.
+ * For details, see https://github.com/LLNL/UnifyCR.
+ * Please read https://github.com/LLNL/UnifyCR/LICENSE for full license text.
+ */
+
+#include "unifycr_const.h"
+#include "unifycr_keyval.h"
+#include "unifycr_log.h"
+
+//#include "config.h"
+
+#if defined(USE_PMIX)
+# include <pmix.h>
+#elif defined(USE_PMI2)
+# include <pmi2.h>
+#else
+# define KEYVAL_USE_FILES 1
+#endif
+
+#include <dirent.h>
+#include <inttypes.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/stat.h>
+
+// UnifyCR keys
+const char* key_runstate        = "unifycr.runstate";
+const char* key_unifycrd_socket = "unifycrd.socket";
+
+// key-value store state
+static int kv_initialized; // = 0
+static int kv_published;   // = 0
+static char kv_myhost[80];
+
+// max supported key and value length
+static size_t kv_max_keylen; // = 0
+static size_t kv_max_vallen; // = 0
+
+#ifndef UNIFYCR_MAX_KV_KEYLEN
+# define UNIFYCR_MAX_KV_KEYLEN 256
+#endif
+
+#ifndef UNIFYCR_MAX_KV_VALLEN
+# define UNIFYCR_MAX_KV_VALLEN 4096
+#endif
+
+#if defined(USE_PMI2)
+
+static int pmi_world_rank;     // = 0
+static int pmi_world_nprocs;   // = 0
+
+static int pmi_errstr_len; // = 0
+static char pmi_errstr[MPI_MAX_ERROR_STRING];
+
+// initialize PMI2
+static int unifycr_pmi2_init(void)
+{
+    int rc, version, nprocs, rank;
+
+    rc = PMI2_Init(NULL, &nprocs, &rank, NULL, &version, NULL);
+    if (rc != MPI_SUCCESS) {
+        MPI_Error_string(rc, pmi_errstr, &pmi_errstr_len);
```

Comment:
Do you mean to use MPI to report a PMI error string?  I don't see an PMI2_Error_string in the pmi2.h, but I'm guessing MPI may not do this portably across different MPI libs.

Reply:
According to PMI2 docs I found, you are supposed to use MPI_Error_string() to get the string version of the PMI error.

Reply:
I ended up removing the use of MPI_Error_string(), and coding up my own function that uses the PMI2 error definitions.